[improvement] Refactor Auditor to simplify the readability

wenbingshen opened the pull request:
### Motivation

Fix #3636 

See explanation in #3636 

### Changes

1. Auditor class: Only reserved to do some necessary initialization work, scheduling of various detection tasks
2. Separate monitoring metrics into `AuditorStats`
3. Separate the auditBookies task to the `AuditorBookieCheckTask`
4. Separate the checkAllLedgers task into the `AuditorCheckAllLedgersTask`
5. Separate placementPolicyCheck task into `AuditorPlacementPolicyCheckTask`
6. Separate replicasCheck to `AuditorReplicasCheckTask`

wenbingshen commented:
rerun failure checks

eolivelli reviewed (approved):
+1
thanks for this improvement, it will help a lot new contributions

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorBookieCheckTask.java:
```diff
@@ -0,0 +1,163 @@
+public class AuditorBookieCheckTask extends AuditorTask {
+    private static final Logger LOG = LoggerFactory.getLogger(AuditorBookieCheckTask.class);
+
+    private BookieLedgerIndexer bookieLedgerIndexer;
```
final?

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:
```diff
@@ -587,7 +270,7 @@ synchronized Future<?> submitAuditTask() {
                 knownBookies.removeAll(bookiesToBeAudited);
                 if (lostBookieRecoveryDelay == 0) {
-                    startAudit(false);
+                    newAuditorBookieCheckTask().startAudit(false);
```
Are there any ways that do not create the `checkTask` instance every time?

wenbingshen replied:
@hangc0276 May be we can use netty Recycler. But I don't think a large number of AuditorTasks will be created in a short period of time, because the inspection tasks are executed periodically, most of them are once a day, and these instances can be recycled by the GC as soon as possible.

If you feel strongly about this, I can add.

wenbingshen replied:
Maybe we can reuse the one AuditorTask, I'll try it.

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorReplicasCheckTask.java:
```diff
@@ -0,0 +1,750 @@
+                    }
+                    List<BookieExpectedToContainSegmentInfo> bookieSegmentInfoList = bookiesSegmentInfoMap
+                            .get(bookieInEnsemble);
+                    if (bookieSegmentInfoList == null) {
```
replace with `computeIfAbsent` ?

wenbingshen replied:
addressed done

wenbingshen commented:
rerun failure checks

wenbingshen commented:
@hangc0276 I have addressed all your comments. PTAL.

dlg99 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorBookieCheckTask.java:
```diff
@@ -0,0 +1,163 @@
+
+    private final BookieLedgerIndexer bookieLedgerIndexer;
+
+    public AuditorBookieCheckTask(Auditor auditor,
```
Can we avoid circular dependency between Auditor and AuditorBookieCheckTask? 
It will backfire eventually.
Most likely you'd notice it if you start adding unit tests/mocks for AuditorBookieCheckTask/other tasks instead of current coarse-leveled test.

wenbingshen replied:
I have addressed.

dlg99 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorCheckAllLedgersTask.java:
```diff
@@ -0,0 +1,240 @@
+    private final Semaphore openLedgerNoRecoverySemaphore;
+    private final int openLedgerNoRecoverySemaphoreWaitTimeoutMSec;
+
+    AuditorCheckAllLedgersTask(Auditor auditor,
```
same, circular dependency

wenbingshen replied:
I have addressed.

dlg99 reviewed (commented):
First of all, thank you for taking on this task.
I recognize it is not simple, given all the iterations of this code and "small changes" snuck there over the years.

Looking at this version I feel it fails a bit short of original plan of refactoring. 
It does split the code into smaller components but it leaves the components interconnected - tasks that have to access Auditor's internals etc. Typically this is a design smell that will backfire in the future when changing the code or it will become a burden when you start writing unit tests and mocks for the tasks
Second point is the unit tests. I think that new tasks deserve some unit tests.

This is a fantastic job so far and I applaud you for doing it.
I hope my comment makes sense and it can be made even better.

wenbingshen commented:
> I hope my comment makes sense and it can be made even better.
@dlg99 Thanks for all your comments. The comments make a lot of sense, I'll try to perfect the whole PR.

wenbingshen commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorTask.java:
```diff
@@ -0,0 +1,114 @@
+
+    protected void waitIfLedgerReplicationDisabled() throws ReplicationException.UnavailableException,
+            InterruptedException {
+        ReplicationEnableCb cb = new ReplicationEnableCb();
```
I have addressed.

wenbingshen commented:
> It does split the code into smaller components but it leaves the components interconnected - tasks that have to access Auditor's internals etc. Typically this is a design smell that will backfire in the future when changing the code or it will become a burden when you start writing unit tests and mocks for the tasks
Second point is the unit tests. I think that new tasks deserve some unit tests.

@dlg99 
1. I updated the code to avoid circular dependency between Auditor and CheckTask.
2. I added unit tests corresponding to different tasks.
3. Now these tasks can run independently without Auditor.
4. These tasks all have basic unit tests, indicating that they can run independently without Auditor. Since there are too many items that can be used for testing, if we want other unit tests to cover more options, I hope to use another PR. The previous unit tests are sufficient to prove that the changes in this PR maintain compatibility.

If tests not fail, PTAL. Thanks!

dlg99 reviewed (approved):
LGTM, thank you!

One test failed, I am re-running the test suite, we'll see if that's a flake.

wenbingshen commented:
I have another bugfix for Auditor. I want to commit after this PR so that I can avoid rebase master. Can you help me to see this PR? Thank you very much. @hangc0276 @dlg99 @eolivelli

hangc0276 commented:
@horizonzy Please help take a look, thanks.

horizonzy commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorPlacementPolicyCheckTask.java:
```diff
@@ -0,0 +1,297 @@
+            placementPolicyCheck();
+            long placementPolicyCheckDuration = stopwatch.stop().elapsed(TimeUnit.MILLISECONDS);
+            int numOfLedgersFoundNotAdheringInPlacementPolicyCheckValue =
+                    auditorStats.getNumOfLedgersFoundNotAdheringInPlacementPolicyCheck().get();
```
Not use auditorStats.getNumOfLedgersFoundNotAdheringInPlacementPolicyCheck, we can define numOfLedgersFoundNotAdheringInPlacementPolicyCheck in AuditorPlacementPolicyCheckTask.
Some other fields also are, just define the guava field in AuditorStats

wenbingshen replied:
I agree! Addressed done.

horizonzy commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:
```diff
@@ -341,155 +153,48 @@ public Auditor(final String bookieIdentifier,
-        this.numLedgersFoundHavingLessThanWQReplicasOfAnEntry = new AtomicInteger(0);
+        this.auditorStats = new AuditorStats(statsLogger);
 
         if (conf.getAuditorMaxNumberOfConcurrentOpenLedgerOperations() <= 0) {
```
Suggest moving the check logic in the AuditorCheckAllLedgersTask construction.

wenbingshen replied:
Addressed done.

horizonzy commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorTask.java:
```diff
@@ -0,0 +1,176 @@
+
+    protected abstract void runTask();
+
+    protected void waitIfLedgerReplicationDisabled() throws ReplicationException.UnavailableException,
```
It only invokes by AuditorBookieCheckTask, move it to AuditorBookieCheckTask.

wenbingshen replied:
Addressed done.

horizonzy commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:
```diff
@@ -341,155 +153,48 @@ public Auditor(final String bookieIdentifier,
         initialize(conf, bkc);
 
+        AuditorTask.ShutdownTaskHandler shutdownTaskHandler = this::submitShutdownTask;
+        AuditorTask.SubmitTaskHandler submitBookieCheckTaskHandler = this::submitBookieCheckTask;
```
It's only for AuditorBookieCheckTask, not define it in the AuditorTask.

horizonzy replied:
I don't think so. If any task needs to do some operations, it can do the operation alone. The SubmitTaskHandler is not common for different tasks.

wenbingshen replied:
I got it. I have submit your feedback. PTAL. Thanks.

horizonzy commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:
```diff
@@ -599,20 +304,20 @@ synchronized Future<?> submitAuditTask() {
                     // if there is no scheduled audit, schedule one
                     auditTask = executor.schedule(() -> {
-                        startAudit(false);
+                        auditorBookieCheckTask.startAudit(false);
```
It's odd. The auditorBookieCheckTask is runnable, here we define a runnable to invoke `auditorBookieCheckTask.startAudit(false)`.

wenbingshen replied:
auditorBookieCheckTask is runnable, and scheduled operation is a complete task execution. The runTask method is also a part of the task. startAudit is only the execution of one of the methods, and does not represent a complete scheduled task.

horizonzy replied:
Ok, that's fine.

horizonzy commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:
```diff
@@ -727,15 +432,19 @@ public void start() {
     }
 
+    protected void submitBookieCheckTask() {
+        executor.submit(auditorBookieCheckTask);
```
Not wrap it, use `executor.submit(auditorBookieCheckTask);` directly.

horizonzy commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:
```diff
@@ -343,167 +154,43 @@ public Auditor(final String bookieIdentifier,
         initialize(conf, bkc);
 
-        executor = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
+        ledgerCheckerExecutor = Executors.newSingleThreadExecutor(new ThreadFactory() {
```
We can define and initialize it in AuditorCheckAllLedgersTask. And add a abstract method `shutdown()` in the AuditorTask, if we need to do some resource clean, we can override the shutdown(). In Auditor shutdown(), to invoke task.shutdown();

wenbingshen replied:
addressed done.

horizonzy commented on bookkeeper-server/src/test/java/org/apache/bookkeeper/replication/AuditorPeriodicCheckTest.java:
```diff
@@ -751,6 +737,24 @@ CountDownLatch getLatch() {
+
+        private static class AuditorTestWrapperTask extends AuditorTask {
+            private final AuditorTask innerTask;
+            private final AtomicReference<CountDownLatch> latchRef;
```
Just confirm, why define latchRef. And the latchRef is shared in all the AuditorTestWrapperTask.

wenbingshen replied:
@horizonzy I think there are several reasons for defining latchRef:
1. Before this PR, the original author has defined latchRef in this test, I just try to ensure that its content does not change and the unit test can pass
2. latchRef is only wrapped in three tasks: auditorCheckAllLedgersTask, auditorPlacementPolicyCheckTask and auditorReplicasCheckTask
3. In the unit test, every time any task is tested, the conf will be changed and the other two tasks will be disabled.
4. LatchRef will countDown() only after the execution of the AuditorTask core logic of the unit test is completed, so that the unit test can then judge the correctness of the metrics in the verification operation and whether it meets expectations.
5. latchRef is used to ensure the execution order among multiple threads
6. If there is no better implementation, we can keep the original author's latchRef implementation

horizonzy commented on bookkeeper-server/src/test/java/org/apache/bookkeeper/replication/AuditorCheckAllLedgersTaskTest.java:
```diff
@@ -0,0 +1,118 @@
+        // 4. verify
+        assertEquals("CHECK_ALL_LEDGERS_TIME", 1, ((TestStatsProvider.TestOpStatsLogger) statsLogger
+                .getOpStatsLogger(ReplicationStats.CHECK_ALL_LEDGERS_TIME)).getSuccessCount());
+        assertEquals("", numLedgers,
```
The first param is an empty string.

wenbingshen replied:
addressed done.

horizonzy commented on bookkeeper-server/src/test/java/org/apache/bookkeeper/replication/AuditorPlacementPolicyCheckTaskTest.java:
```diff
@@ -0,0 +1,102 @@
+                statsLogger.getOpStatsLogger(ReplicationStats.PLACEMENT_POLICY_CHECK_TIME)).getSuccessCount());
+        assertEquals("numOfClosedLedgersAuditedInPlacementPolicyCheck",
+                numLedgers,
+                auditorPlacementPolicyCheckTask.getNumOfClosedLedgersAuditedInPlacementPolicyCheck().get());
```
We can also check `numOfLedgersFoundNotAdheringInPlacementPolicyCheck`.

wenbingshen replied:
addressed done.

horizonzy reviewed (approved):
LGTM.

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorCheckAllLedgersTask.java:
```diff
@@ -0,0 +1,286 @@
+        } catch (ReplicationException.NonRecoverableReplicationException nre) {
+            LOG.error("Non Recoverable Exception while reading from ZK", nre);
+            submitShutdownTask();
+            submitShutdownTask();
```
remove the duplicated `submitShutdownTask()`?

wenbingshen replied:
Good catch. I have fixed it.

hangc0276 reviewed (commented):
Overall LGTM.

wenbingshen commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorStats.java:
```diff
@@ -0,0 +1,294 @@
+    private final AtomicInteger numLedgersHavingLessThanAQReplicasOfAnEntryGuageValue;
+    private final AtomicInteger numLedgersHavingLessThanWQReplicasOfAnEntryGuageValue;
+    private final AtomicInteger underReplicatedLedgersGuageValue;
+    private final StatsLogger statsLogger;
```
Make sense to me. I have submit your feedback. PTAL. Thanks.

wenbingshen replied:
@hangc0276 The statsLogger of auditorStats is used in AuditorTask. After thinking about it, if the task needs to use statsLogger to build bookkeeper internal components, statsLogger is best placed in AuditorStats.

hangc0276 replied:
OK, make sense.

wenbingshen commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java:
```diff
@@ -144,140 +78,18 @@ public class Auditor implements AutoCloseable {
     private final String bookieIdentifier;
-    private volatile Future<?> auditTask;
+    protected volatile Future<?> auditTask;
     private Set<String> bookiesToBeAudited = Sets.newHashSet();
```
Addressed it. PTAL.
